src/compiler/crystal/tools/expand.cr

```diff
@@ -0,0 +1,185 @@
+module Crystal
+  class ExpandResult
+    JSON.mapping({
+      status:     {type: String},
+      message:    {type: String},
+      expansions: {type: Array(ExpandTrace), nilable: true},
+    })
+
+    def initialize(@status, @message)
+    end
+
+    def to_text(io)
+      io.puts message
+      expansions.try &.each_with_index do |expansion, i|
+        io.puts "expansion #{i + 1}:"
+        io << "   "
+        io.puts expansion.original.lines(chomp: false).join "   "
+        io.puts
+        expansion.expandeds.each_with_index do |expanded, j|
+          io << "~> "
+          io.puts expanded.lines(chomp: false).join "   "
+          io.puts
+        end
+      end
+    end
+  end
+
+  class ExpandTrace
+    JSON.mapping({
+      original:  {type: String},
+      expandeds: {type: Array(String)},
+    })
+
+    def initialize(@original, @expandeds)
+    end
+
+    def self.ast_to_s(node)
+      s = String.build { |io| node.to_s(io, emit_doc: true) }
+      return s unless node.is_a?(MacroIf) || node.is_a?(MacroFor)
+      indent = node.location.not_nil!.column_number - 1
```

Comment:
`MacroIf` and `MacroFor` have `MacroBody`, it is just sub-string of source text. For example:

```crystal
begin
  {% if 1 == 1 %}
    hello
  {% end %}
end
```

`MacroIf#to_s` result of `{% if 1 == 1 %}...{% end %}` is:

```crystal
{% if 1 == 1 %}
    hello
  {% end %}
```

So, re-indentation is needed. I'll add specs for it.